include/proxy-wasm/context_interface.h

```diff
@@ -0,0 +1,628 @@
+// Copyright 2016-2019 Envoy Project Authors
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "include/proxy-wasm/compat.h"
+
+#include <time.h>
+#include <atomic>
+#include <chrono>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <vector>
+
+namespace proxy_wasm {
+
+#include "proxy_wasm_common.h"
+#include "proxy_wasm_enums.h"
+
+using Pairs = std::vector<std::pair<string_view, string_view>>;
+using PairsWithStringValues = std::vector<std::pair<string_view, std::string>>;
+using HttpCallToken = uint32_t;
+using GrpcToken = uint32_t;
+using SharedQueueDequeueToken = uint32_t;
+using SharedQueueEnqueueToken = uint32_t;
+
+// TODO: move to SDK.
+enum WasmStreamType {
+  Request,
+  Response,
+  Downstream,
+  Upstream,
+};
+
+// TODO: update SDK.
+enum class ProxyAction : uint32_t {
+  Illegal = 0,
+  Continue = 1,
+  Pause = 2,
+};
+
+struct PluginBase;
+class WasmBase;
+
+/**
+ * BufferInterface provides a interface between proxy-specific buffers and the proxy-independent ABI
+ * implementation. Embedders should subclass BufferInterface to enable the proxy-independent code to
+ * implement ABI calls which use buffers (e.g. the HTTP body).
+ */
+struct BufferInterface {
+  virtual ~BufferInterface() {}
+  virtual size_t size() const = 0;
+  /**
+   * Copy bytes from the buffer into the Wasm VM corresponding to 'wasm'.
+   * @param start is the first buffer byte to copy.
+   * @param length is the length of sequence of buffer bytes to copy.
+   * @param ptr_ptr is the location in the VM address space to place the address of the newly
+   * allocated memory block which contains the copied bytes.
+   * @param size_ptr is the location in the VM address space to place the size of the newly
+   * allocated memory block which contains the copied bytes (i.e. length).
+   * @return a WasmResult with any error or WasmResult::Ok.
+   */
+  virtual WasmResult copyTo(WasmBase *wasm, size_t start, size_t length, uint64_t ptr_ptr,
+                            uint64_t size_ptr) const = 0;
+};
+
+/**
+ * RootGrpcInterface is the interface for gRPC events arriving at RootContext(s).
+ */
+struct RootGrpcInterface {
+  /**
+   * Called on Root Context to with the initial metadata of a grpcStream.
+   * @token is the token returned by grpcStream.
```

Comment:
What is a `grpcStream` here? How come we are using this camel-case for some noun?